### DISCUSSION TITLE
Playing around with an NXP i.MX8M Mini board - Docker with Core/Supervisor/HAOS + Addons?

### DISCUSSION THREAD

**Discussion by ajlennon** [meta-homeassistant/meta-homeassistant#127] [General]
Hi all,

I have a Yocto build with Docker support for one of our NXP i.MX8M Mini boards.

I had a little go with this and it worked incredibly easily!

https://www.influxdata.com/blog/getting-started-home-assistant-docker/

However because I think this is just Core I don't get add-ons and I use quite a lot of these.

So I get that this isn't directly related to meta-homeassistant but is there any easy way to get the supervisor add-ons support via some kind of docker container install or do we need to be using meta-homeassistant to build a base Yocto layer?

![image](https://github.com/user-attachments/assets/04683bbd-bdbd-4e6b-9936-9e10d696e712)

Thanks!

Alex

**Comment by bachp** [comment]
Hi Alex, I think add-ons are a specific thing of [HassOS](https://github.com/home-assistant/operating-system). So supporting them would basically mean to re-create parts of their buildroot environment in Yocto.

While this is not complete impossible to do it is currently out of scope. The goal of this layer is to be able to run Home Assistant as part of a yocto build.

**Comment by moto-timo** [comment]
At least one of the things I wanted a few years ago was Apple HomeKit support. As I recall, there were a number of obvious direct dependencies that we did not have recipes for, plus any nested dependencies that were also not present. For now, I am simply going to be extremely thankful for @Gerry546 's efforts in bringing things into a better state and much more modern than @bachp or I had time to do.

**Reply by Gerry546** [reply]
Thanks @moto-timo , appreciated.

For the question, I have personally never run Supervisor, with addons so what are we exactly talking about. I scrolled through the docs but it just says that it won't work on a Core-only build.

What would essentially be needed?

**Comment by ajlennon** [comment]
Thanks for the comments folks. So just to add in a bit of use-case feedback from me....

You are all more experienced with HA than me and I am still trying to figure things out but I *think* Core is just a docker container that sits on any system, allows integrations but not add-ons (been trying to figure out the difference for ages). Now if I'm right then there is also the Supervisor which is another container (?) which allows management of things such as add-ons and those add-ons live in their own containers. Supervisor can be installed on systems with some fiddling. Then lastly there is HAOS which is the entire OS image they have setup and provide so people don't have to do things themselves.

So it sounds to me as though a meta-homeassistant could be a layer supporting a supervisor install which would give us an ability to install these addons in containers. I do understand this might be out of scope for what people want and I completely understand we don't want people "coming around here" telling other OpenSource people what they should be doing :-D I am just trying to work thoughts out.

Now add-ons is interesting to me as I use HACS which I I had to install myself underneath HA because amongst other things HACS lets me install the Octopus energy integration which I use to monitor my live energy usage. I may be wrong  on this one as maybe HACS can be installed alongside Core and I wouldn't need the add-on support per se (?)

I do find the add-ons really useful though as I run a fairly fully featured HA box with Lets Encrypt, Matter Server, Mosquitto MQTT broker, Grafana and InfluxDB amongst other things

![image](https://github.com/user-attachments/assets/500d949d-9a53-49e1-b69e-31638268c77b)

I am wondering.... and please don't shoot me down in flames I know how it sounds and I've seen the response to somebody else who suggested this on the HA forum. But for some meta-mono CI work I do I have Docker containers nested inside containers. I am wondering if it's possible to build a container with HA in Supervisor install which would then manage nested containers which would give me the "full" HA experience including addons.

Just some thoughts!

**Comment by Gerry546** [comment]
Ah ok, Maybe I can clarify a bit.

Home Assistant Core is the heart of everything. It is Home Assistant, but
in essence it is nothing more than a (rather large) Python Application.

Home Assistant Docker is, I think, taking Core and combining it with all
the prerequisites it needs and combines it into a nice standalone Docker
Image.

Now Supervisor is indeed another step as it sort of works as a Docker
Manager and it can manage and install addons.

And HassOS is one step further, it is Supervisor, but then also on a
tightly controlled and managed custom Linux image.

The key difference between integrations and addons is the following. An
addon is a separate application (mosquito MQTT server, Matter Server)
running on the same OS. Supervisor installs these indeed as Docker
Containers inside the 'Supervisor Swarm'. Now integrations on the other
hand are other Python modules which provide extensions to Core to be able
to interact with external entities. This can be for instance an Addon. But
also an Ikea Tradfri Hub on your home network.

Now what meta-homeassistant currently tries to achieve is to give people
the tool to fully customize their HomeAssistant and run it on any platform
which Yocto supports. HassOS only supports a limited number of hardware
platforms, and Supervisor even less. So you can think of meta-homeassistant
as a custom HassOS. What we do not currently support (yet) is the same way
of enabling people to install, for instance, a Mosquito MQTT server on the
same operating system.

Also, on the use case of meta-homeassistant you would think of integrating
this beforehand (compile time) and not when it is actually already running.
But that idea could change of course

**Comment by ajlennon** [comment]
That's really helpful thanks. So a step to support add-ons would be maybe to integrate a dependency on meta-docker maybe? Or am I on the wrong track?

**Comment by Gerry546** [comment]
No that would be on the right track.

I think the main thing conceptual difference is that meta-homeassistant is not a dockerized thing. So it is just Core, directly running on your own Embedded Linux OS. So the addon idea is having this other application (MQTT-broker for instance) running as well on the same OS.

So for facilitating the same thing as an addon what you could do I think is to facilitate an installation of, say, Mosquitto MQTT on the OS (using [mosquitto.bb](https://layers.openembedded.org/layerindex/recipe/85578/) such that this is running as well. Then you can configure your HA to use this broker.

The big difference is that with SuperVisor you can 'install' addons after you already deployed on the target. Whereas with meta-homeassistatn the OS is more set, so you would do it at compile time using recipes and then fully update the rootFS in one swoop (using RAUC or SWUpdate for example).

That is at least how I am currently running it :)

**Comment by ajlennon** [comment]
Great - am I right in thinking that with HAOS "Core" would be running in a container rather than on the base OS level? I'm thinking what I am trying to get to is to have a Yocto build of something that can host Core/Supervisor/AddOns in as similar a manner as possible to that?
